In [1]:
# Necessary imports
import pandas as pd
import seaborn as sns
import numpy as np
# We want all columns displayed, setting output to 50 to be safe
pd.set_option('display.max_columns',100)

In [2]:
import glob
# Use glob to get all filepaths that match the pattern (*=wildcard)
tmdb_files = sorted(glob.glob("Data/final_tmdb_data*.csv.gz"))
tmdb_files

['Data\\final_tmdb_data_2001.csv.gz',
 'Data\\final_tmdb_data_2002.csv.gz',
 'Data\\final_tmdb_data_2010.csv.gz']

In [3]:
# Use read_csv in a list comprehension and combine with concat to load all files
merged = pd.concat([pd.read_csv(f) for f in tmdb_files] )
merged

imdb_id  adult                     backdrop_path  \
0             0    NaN                               NaN   
1     tt0096056    0.0  /95U3MUDXu4xSCmVLtWgargRipDi.jpg   
2     tt0118926    0.0  /qR3Dk3ctnrrxkAI6I472RhamIbu.jpg   
3     tt0119980    0.0                               NaN   
4     tt0120679    0.0  /s04Ds4xbJU7DzeGVyamccH4LoxF.jpg   
...         ...    ...                               ...   
2734  tt7851834    0.0  /c4x2XPngdiW8Vmc6EmmTXAgzViJ.jpg   
2735  tt8090084    0.0                               NaN   
2736  tt8160720    0.0                               NaN   
2737  tt9164254    1.0                               NaN   
2738  tt9575206    0.0  /yejvp5VvZa6TnXodrQsYUKzF4Cs.jpg   

     belongs_to_collection      budget  \
0                      NaN         NaN   
1                      NaN         0.0   
2                      NaN         0.0   
3                      NaN         0.0   
4                      NaN  12000000.0   
...                    ...         ...   
2734                   NaN         0.0   
2735                   NaN         0.0   
2736                   NaN         0.0   
2737                   NaN         0.0   
2738                   NaN         0.0   

                                                 genres  \
0                                                   NaN   
1                         [{'id': 18, 'name': 'Drama'}]   
2     [{'id': 80, 'name': 'Crime'}, {'id': 18, 'name...   
3     [{'id': 80, 'name': 'Crime'}, {'id': 18, 'name...   
4     [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   
...                                                 ...   
2734                      [{'id': 18, 'name': 'Drama'}]   
2735                     [{'id': 35, 'name': 'Comedy'}]   
2736                                                 []   
2737                                                 []   
2738                      [{'id': 18, 'name': 'Drama'}]   

                                 homepage        id original_language  \
0                                     NaN       NaN               NaN   
1                                     NaN  109809.0                en   
2                                     NaN   20689.0                en   
3                                     NaN  563364.0                en   
4     https://www.miramax.com/movie/frida    1360.0                en   
...                                   ...       ...               ...   
2734                                  NaN  166919.0                ko   
2735                                  NaN  516206.0                en   
2736                                  NaN  273502.0                en   
2737                                  NaN  775304.0                en   
2738                                  NaN  574744.0                fa   

             original_title  \
0                       NaN   
1      Crime and Punishment   
2       The Dancer Upstairs   
3     Random Shooting in LA   
4                     Frida   
...                     ...   
2734                    여의도   
2735          Goodbye Dolly   
2736             Moist Fury   
2737   Text, Lies and Video   
2738                  پنهان   

                                               overview  popularity  \
0                                                   NaN         NaN   
1     A modern day adaptation of Dostoyevsky's class...      10.357   
2     A police detective in a South American country...       8.475   
3     The seamy side of Los Angeles is revealed thro...       0.600   
4     A biography of artist Frida Kahlo, who channel...      15.753   
...                                                 ...         ...   
2734  In bleak Yeouido, crowded by the political wor...       1.624   
2735  Rival suburban divorcees Bettie and Joan must ...       0.600   
2736  It's "Death Wish" meets "Savage Streets" meets...       0.840   
2737  When Andy (Nicco Sky) finds out his hookup wit...       0.000   
2738  Our city has become that much big th

In [4]:
# We should drop the first blank row
merged = merged.drop(0)
merged.head(1)

imdb_id  adult                     backdrop_path belongs_to_collection  \
1  tt0096056    0.0  /95U3MUDXu4xSCmVLtWgargRipDi.jpg                   NaN   

   budget                         genres homepage        id original_language  \
1     0.0  [{'id': 18, 'name': 'Drama'}]      NaN  109809.0                en   

         original_title                                           overview  \
1  Crime and Punishment  A modern day adaptation of Dostoyevsky's class...   

   popularity                       poster_path production_companies  \
1      10.357  /2ckMQwDi11TofiNoaE3sHrYbaCh.jpg                   []   

                                production_countries release_date  revenue  \
1  [{'iso_3166_1': 'US', 'name': 'United States o...   2002-06-01      0.0   

   runtime                                   spoken_languages    status  \
1    126.0  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   

  tagline                 title  video  vote_average  vote_count certification  
1     NaN  Crime and Punishment    0.0         5.385        13.0           NaN

1. How many movies had at least some valid financial information (values > 0 for budget OR revenue)?

In [5]:
# We can use an or operator to make a dataframe with values greater than 0 in budget or revenue
valued = merged[(merged['revenue'] > 0) | (merged['budget'] > 0)]
valued.head(3)

imdb_id  adult                     backdrop_path  \
2  tt0118926    0.0  /qR3Dk3ctnrrxkAI6I472RhamIbu.jpg   
4  tt0120679    0.0  /s04Ds4xbJU7DzeGVyamccH4LoxF.jpg   
5  tt0120804    0.0   /5pGWjnM62Zs0S1xRf3TDL1Xizr.jpg   

                               belongs_to_collection      budget  \
2                                                NaN         0.0   
4                                                NaN  12000000.0   
5  {'id': 17255, 'name': 'Resident Evil Collectio...  33000000.0   

                                              genres  \
2  [{'id': 80, 'name': 'Crime'}, {'id': 18, 'name...   
4  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   
5  [{'id': 27, 'name': 'Horror'}, {'id': 28, 'nam...   

                                           homepage       id  \
2                                               NaN  20689.0   
4               https://www.miramax.com/movie/frida   1360.0   
5  https://www.sonypictures.com/movies/residentevil   1576.0   

  original_language       original_title  \
2                en  The Dancer Upstairs   
4                en                Frida   
5                en        Resident Evil   

                                            overview  popularity  \
2  A police detective in a South American country...       8.475   
4  A biography of artist Frida Kahlo, who channel...      15.753   
5  When a virus leaks from a top-secret facility,...      25.030   

                        poster_path  \
2  /jG662jKzEf63fhcbbN3WiLlz5MX.jpg   
4  /a4hgR6aKoohB6MHni171jbi9BkU.jpg   
5  /1UKNef590A0ZaMnxsscIcWuK1Em.jpg   

                                production_companies  \
2  [{'id': 357, 'logo_path': None, 'name': 'Vía D...   
4  [{'id': 14, 'logo_path': '/m6AHu84oZQxvq7n1rsv...   
5  [{'id': 248, 'logo_path': None, 'name': 'Impac...   

                                production_countries release_date  \
2  [{'iso_3166_1': 'ES', 'name': 'Spain'}, {'iso_...   2002-09-20   
4  [{'iso_3166_1': 'CA', 'name': 'Canada'}, {'iso...   2002-08-29   
5  [{'iso_3166_1': 'CA', 'name': 'Canada'}, {'iso...   2002-03-15   

       revenue  runtime                                   spoken_languages  \
2    5227348.0    132.0  [{'english_name': 'English', 'iso_639_1': 'en'...   
4   56298474.0    123.0  [{'english_name': 'English', 'iso_639_1': 'en'...   
5  102984862.0    100.0  [{'english_name': 'English', 'iso_639_1': 'en'...   

     status                                            tagline  \
2  Released  An honest man caught in a world of intrigue, p...   
4  Released                             Prepare to be seduced.   
5  Released  A secret experiment. A deadly virus. A fatal m...   

                 title  video  vote_average  vote_count certification  
2  The Dancer Upstairs    0.0         6.200        51.0           NaN  
4                Frida    0.0         7.422      1915.0             R  
5        Resident Evil    0.0         6.613      5870.0             R

In [6]:
# With only rows with some budget/revenue in valued dataframe, we just need to count total rows
valued.shape

(3647, 26)

##### There are 3647 movies with some information on budget OR revenue

In [7]:
# We won't want any movies shown without a budget or revenue displayed for future queries
# Using & operator to make new df
df = merged[(merged['revenue'] > 0) & (merged['budget'] > 0)]
df.head(3)

imdb_id  adult                     backdrop_path  \
4  tt0120679    0.0  /s04Ds4xbJU7DzeGVyamccH4LoxF.jpg   
5  tt0120804    0.0   /5pGWjnM62Zs0S1xRf3TDL1Xizr.jpg   
6  tt0120912    0.0  /o1l6OCqA3pYDZHIc5L1s7CEzCMv.jpg   

                               belongs_to_collection       budget  \
4                                                NaN   12000000.0   
5  {'id': 17255, 'name': 'Resident Evil Collectio...   33000000.0   
6  {'id': 86055, 'name': 'Men In Black Collection...  140000000.0   

                                              genres  \
4  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   
5  [{'id': 27, 'name': 'Horror'}, {'id': 28, 'nam...   
6  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   

                                           homepage      id original_language  \
4               https://www.miramax.com/movie/frida  1360.0                en   
5  https://www.sonypictures.com/movies/residentevil  1576.0                en   
6  https://www.sonypictures.com/movies/meninblackii   608.0                en   

    original_title                                           overview  \
4            Frida  A biography of artist Frida Kahlo, who channel...   
5    Resident Evil  When a virus leaks from a top-secret facility,...   
6  Men in Black II  Kay and Jay reunite to provide our best, last ...   

   popularity                       poster_path  \
4      15.753  /a4hgR6aKoohB6MHni171jbi9BkU.jpg   
5      25.030  /1UKNef590A0ZaMnxsscIcWuK1Em.jpg   
6      38.110  /enA22EPyzc2WQ1VVyY7zxresQQr.jpg   

                                production_companies  \
4  [{'id': 14, 'logo_path': '/m6AHu84oZQxvq7n1rsv...   
5  [{'id': 248, 'logo_path': None, 'name': 'Impac...   
6  [{'id': 5, 'logo_path': '/tVoHtTipJDUz5cCUib65...   

                                production_countries release_date  \
4  [{'iso_3166_1': 'CA', 'name': 'Canada'}, {'iso...   2002-08-29   
5  [{'iso_3166_1': 'CA', 'name': 'Canada'}, {'iso...   2002-03-15   
6  [{'iso_3166_1': 'US', 'name': 'United States o...   2002-07-03   

       revenue  runtime                                   spoken_languages  \
4   56298474.0    123.0  [{'english_name': 'English', 'iso_639_1': 'en'...   
5  102984862.0    100.0  [{'english_name': 'English', 'iso_639_1': 'en'...   
6  445135288.0     88.0  [{'english_name': 'English', 'iso_639_1': 'en'...   

     status                                            tagline  \
4  Released                             Prepare to be seduced.   
5  Released  A secret experiment. A deadly virus. A fatal m...   
6  Released                                     Back in black.   

             title  video  vote_average  vote_count certification  
4            Frida    0.0         7.422      1915.0             R  
5    Resident Evil    0.0         6.613      5870.0             R  
6  Men in Black II    0.0         6.400      9331.0         PG-13

2. How many movies are there in each of the certification categories (G/PG/PG-13/R)?

In [8]:
df['certification'].unique()

array(['R', 'PG-13', 'PG', 'G', nan, 'NR'], dtype=object)

In [9]:
# Value counts will pull number counts of each category
df['certification'].value_counts()

R        623
PG-13    555
PG       160
G         47
NR        34
Name: certification, dtype: int64

3. What is the average revenue per certification category?

In [10]:
# A for loop can pull the average revenue per category
list = ['R', 'PG-13', 'PG', 'G', 'NR']
r = 'revenue' # it would seem you can't have quotes within f strings, hence new variable
for i in list:
    new = df[(df['certification'] == i)]
    number = np.mean(new[r]).round(2)
    res = ('{:,}'.format(number))
    print(f'The average revenue for {i} films is: {res}.')

The average revenue for R films is: 44,541,164.55.
The average revenue for PG-13 films is: 111,899,105.99.
The average revenue for PG films is: 198,517,154.04.
The average revenue for G films is: 157,635,444.87.
The average revenue for NR films is: 9,839,150.44.


4. What is the average budget per certification category?

In [11]:
# The process is mostly same as above, replacing revenue with budget
b = 'budget'
for i in list:
    new = df[(df['certification'] == i)]
    number = np.mean(new[b]).round(2)
    res = ('{:,}'.format(number))
    print(f'The average budget for {i} films is: {res}.')

The average budget for R films is: 26,210,286.21.
The average budget for PG-13 films is: 45,238,714.03.
The average budget for PG films is: 58,703,680.75.
The average budget for G films is: 49,829,787.23.
The average budget for NR films is: 7,242,659.68.


In [12]:
# Saving dataframe to folder as csv.gz
df.to_csv('Data/tmdb_results_combined.csv.gz', index=False)

In [13]:
# As we're moving data to sequel, we need its dtypes for a dictionary
from sqlalchemy.types import *

In [14]:
# Create a schema dictionary using SQLAlchemy datatype objects
merged_dtypes_dict = {'imdb_id': CHAR(10), 
                       'revenue': FLOAT(),
                       'budget': FLOAT(),
                       'certification': CHAR(10)}

In [15]:
import pymysql
pymysql.install_as_MySQLdb()
from sqlalchemy import create_engine
from sqlalchemy_utils import create_database, database_exists
from urllib.parse import quote_plus as urlquote
# Create connection string using credentials following this format
# connection = "dialect+driver://username:password@host:port/database"
connection = "mysql+pymysql://root:root@localhost/movies"

In [16]:
engine = create_engine(connection)
database_exists(connection)

True

In [17]:
# create a connection to the database with the engine
conn = engine.connect()

In [18]:
# Making new dataframe to put to sequel with only columns we desire
df_new = df.copy()
df_new = df_new[['imdb_id', 'revenue', 'budget', 'certification']]
df_new.head()

imdb_id      revenue       budget certification
4  tt0120679   56298474.0   12000000.0             R
5  tt0120804  102984862.0   33000000.0             R
6  tt0120912  445135288.0  140000000.0         PG-13
7  tt0121765  649398328.0  120000000.0            PG
8  tt0133240  109578115.0  140000000.0            PG

In [19]:
df_new.to_sql('tmdb_data', conn, dtype = merged_dtypes_dict, if_exists = 'replace', index = False)

1931

In [20]:
q = '''SELECT * FROM tmdb_data
        LIMIT 5;'''
pd.read_sql(q, conn)

imdb_id      revenue       budget certification
0  tt0120679   56298500.0   12000000.0             R
1  tt0120804  102985000.0   33000000.0             R
2  tt0120912  445135000.0  140000000.0         PG-13
3  tt0121765  649398000.0  120000000.0            PG
4  tt0133240  109578000.0  140000000.0            PG

In [21]:
q = '''SELECT * FROM title_basics
        LIMIT 5;'''
pd.read_sql(q, conn)

tconst                                      primary_title  start_year  \
0  tt0035423                                     Kate & Leopold      2001.0   
1  tt0062336  The Tango of the Widower and Its Distorting Mi...      2020.0   
2  tt0069049                         The Other Side of the Wind      2018.0   
3  tt0088751                                  The Naked Monster      2005.0   
4  tt0096056                               Crime and Punishment      2002.0   

   runtime ratings_tconst tmdb_data_imdb_id  
0      118           None              None  
1       70           None              None  
2      122           None              None  
3      100           None              None  
4      126           None              None

In [22]:
# Trying to join on ratings
q = '''SELECT title_basics.primary_title, ratings.average_rating FROM title_basics
        JOIN ratings ON ratings.tconst = title_basics.ratings_tconst'''
pd.read_sql(q,conn)

Empty DataFrame
Columns: [primary_title, average_rating]
Index: []

In [23]:
# How many movies are there in each of the certification categories (G/PG/PG-13/R)?


In [24]:
# What is the average revenue per certification category?


In [25]:
# What is the average budget per certification category?
